StringImpl::copyCharacters incorrectly uses memcpy on destination

darinadler opened the pull request:
#### 71265755b78a83c1cc5614e3fede36899e4e3a05
<pre>
StringImpl::copyCharacters incorrectly uses memcpy on destination
pointers that may be null
<a href="https://bugs.webkit.org/show_bug.cgi?id=246260">https://bugs.webkit.org/show_bug.cgi?id=246260</a>

We should assert StringImpl::copyCharacters does not discard information
<a href="https://bugs.webkit.org/show_bug.cgi?id=205355">https://bugs.webkit.org/show_bug.cgi?id=205355</a>

rdar://problem/100962334

Reviewed by Yusuke Suzuki.

After studying the call sites of StringImpl::copyCharacters and its
implementation, it is clear that many callers can pass nullptr for either
the source or destination when the length passed in is 0. Since
std::memcpy behavior is undefined in such cases, need to change the
implementation of StringImpl::copyCharacters to not use it in those
cases. In the end, decided to check for zero length even though I advised
against that previously.

Visiting all the call sites of StringImpl::copyCharacters led to
discovering some other anomalies that are fixed with improved coding
idioms, no major changes.

* Source/JavaScriptCore/API/OpaqueJSString.cpp:
(OpaqueJSString::characters): Updated for name change of
StringView::getCharacters.

* Source/JavaScriptCore/jit/ExecutableAllocationFuzz.cpp: Added include
of NeverDestroyed.h.

* Source/JavaScriptCore/parser/Lexer.cpp:
(JSC::Lexer&lt;T&gt;::parseCommentDirectiveValue): Use String::make8Bit instead
of StringImpl::createUninitialized, for code that is more straightforward.
Also wondering why we don&apos;t need this optimization in more places.

* Source/JavaScriptCore/runtime/IntlLocale.cpp:
(JSC::LocaleIDBuilder::setKeywordValue): Use StringView::getCharacters.

* Source/JavaScriptCore/runtime/IntlObjectInlines.h:
(JSC::ListFormatInput::ListFormatInput): Use String::convertTo16Bit. This
eliminates the need for m_retainedUpconvertedStrings.

* Source/JavaScriptCore/runtime/JSString.cpp:
(JSC::JSRopeString::resolveRopeInternal const): Tweaked the comment
wording for clarity. Use StringView::getCharacters.
(JSC::JSRopeString::resolveRopeInternalNoSubstring const): Ditto.
(JSC::JSRopeString::resolveRopeSlowCase const): Ditto.

* Source/JavaScriptCore/runtime/JSStringJoiner.cpp:
(JSC::appendStringToData): Updated for name change of
StringView::getCharacters.

* Source/JavaScriptCore/runtime/StringPrototype.cpp:
(JSC::jsSpliceSubstrings): Removed unneeded checks for zero length;
adding a branch to optimize that case isn&apos;t valuable and it&apos;s not needed
for correctness.
* Source/JavaScriptCore/runtime/StringPrototypeInlines.h:
(JSC::jsSpliceSubstringsWithSeparators): Ditto. Also use
StringView::getCharacters.

* Source/JavaScriptCore/tools/JSDollarVM.cpp:
(JSC::functionMake16BitStringIfPossible): Use String::convertTo16Bit.

* Source/JavaScriptCore/jit/ExecutableAllocationFuzz.cpp: Added include
of NeverDestroyed.h.

* Source/WTF/wtf/DateMath.cpp:
(WTF::validateTimeZone): Use StringView::upconvertedCharacters.

* Source/WTF/wtf/LogInitialization.cpp: Added include of NeverDestroyed.h.
* Source/WTF/wtf/Logger.cpp: Ditto.

* Source/WTF/wtf/SortedArrayMap.h: Added support for a case-sensitive
version of PackedASCIILowerCodes, named PackedASCIILiteral, and for using
a span of characters as the key in the &quot;packed&quot; case without first
converting it to a StringView and then having to check is8Bit at runtime.

* Source/WTF/wtf/Threading.cpp: Added include of NeverDestroyed.h.
* Source/WTF/wtf/generic/RunLoopGeneric.cpp: Ditto.
* Source/WTF/wtf/linux/RealTimeThreads.cpp: Ditto.

* Source/WTF/wtf/text/ASCIIFastPath.h:
(WTF::copyLCharsFromUCharSource): Deleted. Moved into
StringImpl::copyCharacters.

* Source/WTF/wtf/text/AtomString.h: Removed unneeded includes and the
constructor that takes a Vector. Removed unneeded explicit
implementations of copy and move constructors and assignment operators
that are the same as the defaults. Added a lookUp function.

* Source/WTF/wtf/text/Base64.h: Removed the overloads that take a String
so we don&apos;t implicitly create a String when a StringView will do.

* Source/WTF/wtf/text/StringConcatenate.h: Updated for name change of
StringView::getCharacters.

* Source/WTF/wtf/text/StringImpl.h:
(WTF::StringImpl::copyCharacters): Used overloading instead of
unnecessarily complex template programming. Made the version that uses
memcpy do a zero-length check so we aren&apos;t relying on undefined behavior
of memcpy. Moved the code from copyLCharsFromUCharSource here.
(WTF::StringImpl::removeCharactersImpl): Removed unneeded zero-length
check before calling StringImpl::copyCharacters.

* Source/WTF/wtf/text/StringView.h: Renamed getCharactersWithUpconvert to
just getCharacters since it can be used in many cases that don&apos;t involve
upconversion. Also implemented it as a template since there seems no
disadvantage to doing so. Added getCharacters8, getCharacters16, span8,
and span16.

* Source/WTF/wtf/text/IntegerToStringConversion.h: Added a missing
include.

* Source/WTF/wtf/text/WTFString.cpp:
(WTF::String::charactersWithoutNullTermination const): Removed unnecessary
use of size_t.
(WTF::String::make8Bit): Renamed from make8BitFrom16BitSource, changed the
argument type from size_t to unsigned. Changed to return an empty string
rather than a null string by removing a special case for zero length.
(WTF::String::convertTo16Bit): Added. Replaces make16BitFrom8BitSource,
since all callers can do better by converting a String in place, and we
can significantly simplify them since this efficiently does nothing if
characters are already 16-bit.
(WTF::fromUTF8Impl): Rewrote the &quot;crash if size is too big&quot; to use
RELEASE_ASSERT for clarity.
(WTF::String::fromUTF8WithLatin1Fallback): Added a missing RELEASE_ASSERT
for the fallback case, just like the one above.

* Source/WTF/wtf/text/WTFString.h: Updated for the above changes. Removed
some stray semicolons. Removed the constructor that takes a Vector.

* Source/WTF/wtf/text/icu/UTextProviderLatin1.cpp:
(WTF::uTextLatin1Access): Removed unneeded static_cast.
(WTF::uTextLatin1Extract): Ditto.
(WTF::textLatin1ContextAwareMoveInPrimaryContext): Ditto.

* Source/WebCore/Modules/websockets/ThreadableWebSocketChannelClientWrapper.cpp:
(WebCore::ThreadableWebSocketChannelClientWrapper::setSubprotocol):
Updated for name change of StringView::getCharacters.
(WebCore::ThreadableWebSocketChannelClientWrapper::setExtensions): Ditto.
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::normalizeSpaces): Ditto.

* Source/WebCore/html/HTMLImageElement.cpp:
(WebCore::HTMLImageElement::hasLazyLoadableAttributeValue): Take a
StringView.
* Source/WebCore/html/HTMLImageElement.h: Updated for the above.

* Source/WebCore/html/LinkRelAttribute.cpp:
(WebCore::LinkRelAttribute::LinkRelAttribute): Moved most initialization
into the class definition. Take a StringView, and don&apos;t allocate any
String as part of parsing.
* Source/WebCore/html/LinkRelAttribute.h: Updated for the above.

* Source/WebCore/html/parser/AtomHTMLToken.h: Updated to use
String::make8Bit. Removed the overload of make8BitFrom16BitSource that
took a Vector directly since this was the only place it was used, and such
functions bloat the String interface with little benefit.

* Source/WebCore/html/parser/HTMLMetaCharsetParser.cpp:
(WebCore::extractCharset): Take StringView.
(WebCore::HTMLMetaCharsetParser::processMeta): Pass StringView to
encodingFromMetaAttributes.
(WebCore::HTMLMetaCharsetParser::encodingFromMetaAttributes): Take
StringView.
(WebCore::HTMLMetaCharsetParser::checkForMetaCharset): Use
AtomString::lookUp since we only need to handle existing known tags, not
any that aren&apos;t already in the atom string table.
* Source/WebCore/html/parser/HTMLMetaCharsetParser.h: Update
encodingFromMetaAttributes to take StringView.

* Source/WebCore/html/parser/HTMLPreloadScanner.cpp:
(WebCore::TokenPreloadScanner::tagIdFor): Use a sorted array map instead
of contructing an atom string.
(WebCore::TokenPreloadScanner::StartTagScanner::processAttributes): Use
AtomString::lookUp and StringView.
(WebCore::TokenPreloadScanner::StartTagScanner::processImageAndScriptAttribute):
Use StringView for attribute values, making strings only when the
attribute name indicates we should store the value.
(WebCore::TokenPreloadScanner::StartTagScanner::processAttribute): Ditto.
(WebCore::TokenPreloadScanner::StartTagScanner::setURLToLoad): Take a
StringView, also use a separate named function instead of a mysterious
boolean argument.
(WebCore::TokenPreloadScanner::StartTagScanner::setURLToLoadAllowingReplacement):
More of the same.
(WebCore::HTMLPreloadScanner::scan): Use AtomString::lookUp since we only
need to handle existing known tags, not any that aren&apos;t already in the
atom string table.

* Source/WebCore/html/parser/HTMLSrcsetParser.cpp:
(WebCore::bestFitSourceForImageAttributes): Take StringView.
* Source/WebCore/html/parser/HTMLSrcsetParser.h: Update for the above.

* Source/WebCore/html/parser/HTMLTreeBuilder.cpp:
(WebCore::HTMLTreeBuilder::ExternalCharacterTokenBuffer::makeString const):
Updated for name change of String::make8Bit.

* Source/WebCore/layout/formattingContexts/inline/InlineLineBuilder.cpp:
(WebCore::Layout::endsWithSoftWrapOpportunity): Use
String::convertTo16Bit.

* Source/WebCore/platform/glib/ApplicationGLib.cpp: Added include of
NeverDestroyed.h.

* Source/WebCore/platform/graphics/ComplexTextController.cpp:
(WebCore::ComplexTextController::collectComplexTextRuns): Use
String::convertTo16Bit.

* Source/WebCore/page/PageSerializer.cpp:
(WebCore::isCharsetSpecifyingNode): Handle the type of the argument in a
simpler way, and use HTMLMetaElement instead of metaTag. Use StringView
instead of atom strings for encodingFromMetaAttributes.

* Source/WebCore/platform/graphics/StringTruncator.cpp:
(WebCore::centerTruncateToBuffer): Updated for name change of
StringView::getCharacters.
(WebCore::rightTruncateToBuffer): Ditto.
(WebCore::rightClipToCharacterBuffer): Ditto.
(WebCore::rightClipToWordBuffer): Ditto.
(WebCore::leftTruncateToBuffer): Ditto.
(WebCore::truncateString): Ditto.

* Source/WebCore/platform/graphics/TextRun.h: Added textAsString.
There&apos;s one new call site that takes advantage of taking the string out.
We could also just consider changing the text funtion to return a
const String&amp;, but to do that safely we&apos;d have to make sure no callers
are using substring or similar functions.

* Source/WebCore/platform/graphics/gstreamer/eme/GStreamerEMEUtilities.h:
(WebCore::GStreamerEMEUtilities::uuidToKeySystem): Removed the peculiar
use of NeverDestroyed here. Since ASCIILiteral is just a pointer, used
it as a return type instead of const ASCIILiteral&amp;.

* Source/WebCore/platform/graphics/win/FontCacheWin.cpp:
(WebCore::FontCache::systemFallbackForCharacters): Updated for name change
of StringView::getCharacters. Would be more elegant to add a
platform-specific StringView::getWideCharactersWithNullTermination
function since this pattern repeats.
(WebCore::createGDIFont): Ditto.
(WebCore::FontCache::getFontSelectionCapabilitiesInFamily): Ditto.

* Source/WebCore/platform/mediastream/CaptureDevice.h: Added include of
NeverDestroyed.h.

* Source/WebCore/platform/win/ClipboardUtilitiesWin.cpp:
(WebCore::createGlobalData): Updated for name change of
StringView::getCharacters.
* Source/WebCore/platform/win/PasteboardWin.cpp:
(WebCore::fileSystemPathFromURLOrTitle): Ditto.
(WebCore::Pasteboard::writeURLToDataObject): Ditto.
(WebCore::createGlobalImageFileDescriptor): Ditto.

* Source/WebKitLegacy/win/DOMCoreClasses.cpp:
(DOMElement::font): Updated for name change of StringView::getCharacters.
Would be more elegant to add a platform-specific StringView::getWideCharacters
functionsince this pattern repeats.

* Source/WebCore/platform/xr/openxr/OpenXRInstance.cpp: Added include of
NeverDestroyed.h.
* Source/WebKit/Platform/LogInitialization.cpp: Ditto.

* Source/WebKit/Shared/API/c/WKString.cpp:
(WKStringGetCharacters): Updated for name change of
StringView::getCharacters.

* Source/WebKit/UIProcess/Inspector/mac/WebInspectorUIProxyMac.mm:
(WebKit::WebInspectorUIProxy::showSavePanel): Make the String to pass to
base64Decode explicitly from the NSString.

* Source/WebKit/webpushd/PushService.mm:
(WebPushD::base64URLDecode): Added. Makes the String to pass explicitly
from the NSString passed in.
(WebPushD::base64Decode): Ditto.

* Source/WebKitLegacy/win/WebKitGraphics.cpp:
(CenterTruncateStringToWidth): Ditto.
(RightTruncateStringToWidth): Ditto.

* Source/WebKitLegacy/win/WebView.cpp:
(WebView::onIMERequestReconvertString): Ditto.

* Tools/WebKitTestRunner/InjectedBundle/InjectedBundlePage.cpp: Added
include of NeverDestroyed.h.

Canonical link: <a href="https://commits.webkit.org/255600@main">https://commits.webkit.org/255600@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0e78144a461edcddcf803f4e21b9ee698af97c90

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/93050 "6 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/2252 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/23642 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/102761 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/162984 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/97052 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/2262 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/30579 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/85432 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/98875 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/98716 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/1556 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/79524 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/28470 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/83470 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/83195 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/71582 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/84371 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/36973 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/17099 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/38/builds/79439 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/34786 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/18309 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/38/builds/79439 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3877 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/38658 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/40890 "Passed tests") | [  ~~🛠 jsc-mips~~](https://ews-build.webkit.org/#/builders/37/builds/82072 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/40584 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/37500 "Passed tests") | [  ~~🧪 jsc-mips-tests~~](https://ews-build.webkit.org/#/builders/37/builds/82072 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/runtime/StringPrototypeInlines.h:
```diff
@@ -127,16 +125,14 @@ ALWAYS_INLINE JSString* jsSpliceSubstringsWithSeparators(JSGlobalObject* globalO
-                }
+                auto substring = StringView { source }.substring(substringRanges[i].begin(), substringRanges[i].distance());
+                substring.getCharacters(buffer + bufferPos.value());
+                bufferPos += substring.length();
```
Is it a bit wasteful since we already know that this source is 8Bit?

darinadler replied:
If you think that’s a critical optimization, I can make a `getCharacters8` that asserts the string is 8-bit and skips the check. I would prefer that to writing out the `copyCharacters`.

Constellation commented on Source/JavaScriptCore/runtime/StringPrototypeInlines.h:
```diff
@@ -127,16 +125,14 @@ ALWAYS_INLINE JSString* jsSpliceSubstringsWithSeparators(JSGlobalObject* globalO
-                }
+                StringView separator = separators[i];
+                separator.getCharacters(buffer + bufferPos.value());
+                bufferPos += separator.length();
```
Ditto.

Constellation commented on Source/WTF/wtf/text/WTFString.cpp:
```diff
@@ -470,37 +470,29 @@ CString String::utf8(ConversionMode mode) const
 {
-    if (length > StringImplShape::MaxLength)
-        CRASH();
+    RELEASE_ASSERT(isInBounds<unsigned>(length));
```
`MaxLength` is INT32_MAX, and this is important to make all `String.length` JS Int32 (makes DFG / FTL optimization work well). So, let's use `isInBounds<int32_t>(length)`.

darinadler replied:
I’m happy to make the change. I will do `RELEASE_ASSERT(length <= String::MaxLength)`. But please note that check was not lost even in my patch; `StringImpl::create` enforces that maximum. This outers assertion was just checking that the `length` won’t be chopped down from `size_t` to `unsigned` even before we get inside that function.

Constellation commented on Source/WTF/wtf/text/WTFString.cpp:
```diff
@@ -550,8 +542,10 @@ String String::fromUTF8(const CString& s)
     String utf8 = fromUTF8(string, size);
-    if (!utf8)
+    if (!utf8) {
+        RELEASE_ASSERT(isInBounds<unsigned>(size));
```
Ditto.

darinadler replied:
Same here, the `String` constructor enforces the maximum. But as with the other case, I will write the stronger check as you request.

Constellation commented on Source/WebCore/html/parser/HTMLPreloadScanner.cpp:
```diff
@@ -53,26 +53,26 @@ using namespace HTMLNames;
 {
-    AtomString tagName(data);
-    if (tagName == imgTag)
+    auto knownTagName = AtomString::lookUp(data.data(), data.size());
```
If this lookUp returns null, then, it means that this data is not a known one. So how about,

```
if (knownTagName.isNull())
    return TagId::Unknown;
```

darinadler replied:
New version uses `SortedArrayMap` here instead.

Constellation commented on Source/WebCore/html/parser/HTMLMetaCharsetParser.cpp:
```diff
@@ -157,21 +156,21 @@ bool HTMLMetaCharsetParser::checkForMetaCharset(const char* data, size_t length)
         bool isEnd = token->type() == HTMLToken::Type::EndTag;
         if (isEnd || token->type() == HTMLToken::Type::StartTag) {
-            AtomString tagName(token->name());
+            auto knownTagName = AtomString::lookUp(token->name().data(), token->name().size());
```
Ditto.

darinadler replied:
OK.

Constellation commented on Source/WebCore/html/parser/HTMLPreloadScanner.cpp:
```diff
@@ -515,7 +519,7 @@ void HTMLPreloadScanner::scan(HTMLResourcePreloader& preloader, Document& docume
     while (auto token = m_tokenizer.nextToken(m_source)) {
         if (token->type() == HTMLToken::Type::StartTag)
-            m_tokenizer.updateStateFor(AtomString(token->name()));
+            m_tokenizer.updateStateFor(AtomString::lookUp(token->name().data(), token->name().size()));
```
`lookUp` is different from normal constructor since it can return null AtomString when we do not find the corresponding one in the table (so, it is just `lookUp`).

darinadler replied:
Yes. The `updateStateFor` function only compares against known attribute names. So if it’s not already an existing atom string, it’s fine to pass null since it’s known to not be any of the known attribute names. I could also check for null and not call `updateStateFor` at all in that case.

darinadler replied:
I’m going to leave this as-is.

Constellation commented on Source/WebCore/html/parser/HTMLPreloadScanner.cpp:
```diff
@@ -123,26 +123,25 @@ class TokenPreloadScanner::StartTagScanner {
-            AtomString attributeName(attribute.name);
-            AtomString attributeValue(attribute.value);
-            processAttribute(WTFMove(attributeName), WTFMove(attributeValue), pictureState);
+            auto knownAttributeName = AtomString::lookUp(attribute.name.data(), attribute.name.size());
```
It can become null AtomString, is it OK?

darinadler replied:
Yes, if it’s null then it won’t be equal to any of the atom string constants we compare it against. That’s all we ever do with this name, compare it with constants.

Ms2ger commented on Source/WTF/wtf/text/StringImpl.h:
```diff
@@ -1141,54 +1143,126 @@ inline void StringImpl::deref()
 
-template<typename SourceCharacterType, typename DestinationCharacterType>
-inline void StringImpl::copyCharacters(DestinationCharacterType* destination, const SourceCharacterType* source, unsigned numCharacters)
+template<typename CharecterType> inline void StringImpl::copyCharacters(CharecterType* destination, const CharecterType* source, unsigned length)
```
Nit: CharecterType → CharacterType

darinadler replied:
Fixed.

Constellation commented on Source/JavaScriptCore/parser/Lexer.cpp:
```diff
@@ -1843,31 +1843,24 @@ template <typename T>
-        if (!isLatin1(m_current))
-            hasNonLatin1 = true;
+        if constexpr (std::is_same_v<T, UChar>)
+            containsOnlyLatin1 &= isLatin1(m_current);
```
How about getting merged UChar and finally performing `isLatin1(merged)` once?
```
if constexpr (std::is_same_v<T, UChar>)
    merged |= m_current;
```
and
```
bool containsOnlyLatin1 = true;
if constexpr (std::is_same_v<T, UChar>)
    containsOnlyLatin1 = isLatin1(merged);
```

darinadler replied:
Sure, did that.

darinadler replied:
I _almost_ factored the code from `String::isAllLatin1` out into a helper that takes a span so I could use it here. These days I think such algorithms should work on `Span<LChar>`, `Span<UChar>` and `StringView`, and typically I would not add them as member functions to `String`, `AtomString`, or `StringImpl`. The old functions that start with the word characters that take pointers and lengths probably need to be changed to take `Span` instead.

Constellation reviewed (approved):
r=me. Since this is touching one of the hottest path, please monitor JetStream2 / Speedometer2 data to ensure that we do not have regression.

darinadler committed:
StringImpl::copyCharacters incorrectly uses memcpy on destination
pointers that may be null
https://bugs.webkit.org/show_bug.cgi?id=246260

We should assert StringImpl::copyCharacters does not discard information
https://bugs.webkit.org/show_bug.cgi?id=205355

rdar://problem/100962334

Reviewed by Yusuke Suzuki.

After studying the call sites of StringImpl::copyCharacters and its
implementation, it is clear that many callers can pass nullptr for either
the source or destination when the length passed in is 0. Since
std::memcpy behavior is undefined in such cases, need to change the
implementation of StringImpl::copyCharacters to not use it in those
cases. In the end, decided to check for zero length even though I advised
against that previously.

Visiting all the call sites of StringImpl::copyCharacters led to
discovering some other anomalies that are fixed with improved coding
idioms, no major changes.

* Source/JavaScriptCore/API/OpaqueJSString.cpp:
(OpaqueJSString::characters): Updated for name change of
StringView::getCharacters.

* Source/JavaScriptCore/jit/ExecutableAllocationFuzz.cpp: Added include
of NeverDestroyed.h.

* Source/JavaScriptCore/parser/Lexer.cpp:
(JSC::Lexer<T>::parseCommentDirectiveValue): Use String::make8Bit instead
of StringImpl::createUninitialized, for code that is more straightforward.
Also wondering why we don't need this optimization in more places.

* Source/JavaScriptCore/runtime/IntlLocale.cpp:
(JSC::LocaleIDBuilder::setKeywordValue): Use StringView::getCharacters.

* Source/JavaScriptCore/runtime/IntlObjectInlines.h:
(JSC::ListFormatInput::ListFormatInput): Use String::convertTo16Bit. This
eliminates the need for m_retainedUpconvertedStrings.

* Source/JavaScriptCore/runtime/JSString.cpp:
(JSC::JSRopeString::resolveRopeInternal const): Tweaked the comment
wording for clarity. Use StringView::getCharacters.
(JSC::JSRopeString::resolveRopeInternalNoSubstring const): Ditto.
(JSC::JSRopeString::resolveRopeSlowCase const): Ditto.

* Source/JavaScriptCore/runtime/JSStringJoiner.cpp:
(JSC::appendStringToData): Updated for name change of
StringView::getCharacters.

* Source/JavaScriptCore/runtime/StringPrototype.cpp:
(JSC::jsSpliceSubstrings): Removed unneeded checks for zero length;
adding a branch to optimize that case isn't valuable and it's not needed
for correctness.
* Source/JavaScriptCore/runtime/StringPrototypeInlines.h:
(JSC::jsSpliceSubstringsWithSeparators): Ditto. Also use
StringView::getCharacters.

* Source/JavaScriptCore/tools/JSDollarVM.cpp:
(JSC::functionMake16BitStringIfPossible): Use String::convertTo16Bit.

* Source/JavaScriptCore/jit/ExecutableAllocationFuzz.cpp: Added include
of NeverDestroyed.h.

* Source/WTF/wtf/DateMath.cpp:
(WTF::validateTimeZone): Use StringView::upconvertedCharacters.

* Source/WTF/wtf/LogInitialization.cpp: Added include of NeverDestroyed.h.
* Source/WTF/wtf/Logger.cpp: Ditto.

* Source/WTF/wtf/SortedArrayMap.h: Added support for a case-sensitive
version of PackedASCIILowerCodes, named PackedASCIILiteral, and for using
a span of characters as the key in the "packed" case without first
converting it to a StringView and then having to check is8Bit at runtime.

* Source/WTF/wtf/Threading.cpp: Added include of NeverDestroyed.h.
* Source/WTF/wtf/generic/RunLoopGeneric.cpp: Ditto.
* Source/WTF/wtf/linux/RealTimeThreads.cpp: Ditto.

* Source/WTF/wtf/text/ASCIIFastPath.h:
(WTF::copyLCharsFromUCharSource): Deleted. Moved into
StringImpl::copyCharacters.

* Source/WTF/wtf/text/AtomString.h: Removed unneeded includes and the
constructor that takes a Vector. Removed unneeded explicit
implementations of copy and move constructors and assignment operators
that are the same as the defaults. Added a lookUp function.

* Source/WTF/wtf/text/Base64.h: Removed the overloads that take a String
so we don't implicitly create a String when a StringView will do.

* Source/WTF/wtf/text/StringConcatenate.h: Updated for name change of
StringView::getCharacters.

* Source/WTF/wtf/text/StringImpl.h:
(WTF::StringImpl::copyCharacters): Used overloading instead of
unnecessarily complex template programming. Made the version that uses
memcpy do a zero-length check so we aren't relying on undefined behavior
of memcpy. Moved the code from copyLCharsFromUCharSource here.
(WTF::StringImpl::removeCharactersImpl): Removed unneeded zero-length
check before calling StringImpl::copyCharacters.

* Source/WTF/wtf/text/StringView.h: Renamed getCharactersWithUpconvert to
just getCharacters since it can be used in many cases that don't involve
upconversion. Also implemented it as a template since there seems no
disadvantage to doing so. Added getCharacters8, getCharacters16, span8,
and span16.

* Source/WTF/wtf/text/IntegerToStringConversion.h: Added a missing
include.

* Source/WTF/wtf/text/WTFString.cpp:
(WTF::String::charactersWithoutNullTermination const): Removed unnecessary
use of size_t.
(WTF::String::make8Bit): Renamed from make8BitFrom16BitSource, changed the
argument type from size_t to unsigned. Changed to return an empty string
rather than a null string by removing a special case for zero length.
(WTF::String::convertTo16Bit): Added. Replaces make16BitFrom8BitSource,
since all callers can do better by converting a String in place, and we
can significantly simplify them since this efficiently does nothing if
characters are already 16-bit.
(WTF::fromUTF8Impl): Rewrote the "crash if size is too big" to use
RELEASE_ASSERT for clarity.
(WTF::String::fromUTF8WithLatin1Fallback): Added a missing RELEASE_ASSERT
for the fallback case, just like the one above.

* Source/WTF/wtf/text/WTFString.h: Updated for the above changes. Removed
some stray semicolons. Removed the constructor that takes a Vector.

* Source/WTF/wtf/text/icu/UTextProviderLatin1.cpp:
(WTF::uTextLatin1Access): Removed unneeded static_cast.
(WTF::uTextLatin1Extract): Ditto.
(WTF::textLatin1ContextAwareMoveInPrimaryContext): Ditto.

* Source/WebCore/Modules/websockets/ThreadableWebSocketChannelClientWrapper.cpp:
(WebCore::ThreadableWebSocketChannelClientWrapper::setSubprotocol):
Updated for name change of StringView::getCharacters.
(WebCore::ThreadableWebSocketChannelClientWrapper::setExtensions): Ditto.
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::normalizeSpaces): Ditto.

* Source/WebCore/html/HTMLImageElement.cpp:
(WebCore::HTMLImageElement::hasLazyLoadableAttributeValue): Take a
StringView.
* Source/WebCore/html/HTMLImageElement.h: Updated for the above.

* Source/WebCore/html/LinkRelAttribute.cpp:
(WebCore::LinkRelAttribute::LinkRelAttribute): Moved most initialization
into the class definition. Take a StringView, and don't allocate any
String as part of parsing.
* Source/WebCore/html/LinkRelAttribute.h: Updated for the above.

* Source/WebCore/html/parser/AtomHTMLToken.h: Updated to use
String::make8Bit. Removed the overload of make8BitFrom16BitSource that
took a Vector directly since this was the only place it was used, and such
functions bloat the String interface with little benefit.

* Source/WebCore/html/parser/HTMLMetaCharsetParser.cpp:
(WebCore::extractCharset): Take StringView.
(WebCore::HTMLMetaCharsetParser::processMeta): Pass StringView to
encodingFromMetaAttributes.
(WebCore::HTMLMetaCharsetParser::encodingFromMetaAttributes): Take
StringView.
(WebCore::HTMLMetaCharsetParser::checkForMetaCharset): Use
AtomString::lookUp since we only need to handle existing known tags, not
any that aren't already in the atom string table.
* Source/WebCore/html/parser/HTMLMetaCharsetParser.h: Update
encodingFromMetaAttributes to take StringView.

* Source/WebCore/html/parser/HTMLPreloadScanner.cpp:
(WebCore::TokenPreloadScanner::tagIdFor): Use a sorted array map instead
of contructing an atom string.
(WebCore::TokenPreloadScanner::StartTagScanner::processAttributes): Use
AtomString::lookUp and StringView.
(WebCore::TokenPreloadScanner::StartTagScanner::processImageAndScriptAttribute):
Use StringView for attribute values, making strings only when the
attribute name indicates we should store the value.
(WebCore::TokenPreloadScanner::StartTagScanner::processAttribute): Ditto.
(WebCore::TokenPreloadScanner::StartTagScanner::setURLToLoad): Take a
StringView, also use a separate named function instead of a mysterious
boolean argument.
(WebCore::TokenPreloadScanner::StartTagScanner::setURLToLoadAllowingReplacement):
More of the same.
(WebCore::HTMLPreloadScanner::scan): Use AtomString::lookUp since we only
need to handle existing known tags, not any that aren't already in the
atom string table.

* Source/WebCore/html/parser/HTMLSrcsetParser.cpp:
(WebCore::bestFitSourceForImageAttributes): Take StringView.
* Source/WebCore/html/parser/HTMLSrcsetParser.h: Update for the above.

* Source/WebCore/html/parser/HTMLTreeBuilder.cpp:
(WebCore::HTMLTreeBuilder::ExternalCharacterTokenBuffer::makeString const):
Updated for name change of String::make8Bit.

* Source/WebCore/layout/formattingContexts/inline/InlineLineBuilder.cpp:
(WebCore::Layout::endsWithSoftWrapOpportunity): Use
String::convertTo16Bit.

* Source/WebCore/platform/glib/ApplicationGLib.cpp: Added include of
NeverDestroyed.h.

* Source/WebCore/platform/graphics/ComplexTextController.cpp:
(WebCore::ComplexTextController::collectComplexTextRuns): Use
String::convertTo16Bit.

* Source/WebCore/page/PageSerializer.cpp:
(WebCore::isCharsetSpecifyingNode): Handle the type of the argument in a
simpler way, and use HTMLMetaElement instead of metaTag. Use StringView
instead of atom strings for encodingFromMetaAttributes.

* Source/WebCore/platform/graphics/StringTruncator.cpp:
(WebCore::centerTruncateToBuffer): Updated for name change of
StringView::getCharacters.
(WebCore::rightTruncateToBuffer): Ditto.
(WebCore::rightClipToCharacterBuffer): Ditto.
(WebCore::rightClipToWordBuffer): Ditto.
(WebCore::leftTruncateToBuffer): Ditto.
(WebCore::truncateString): Ditto.

* Source/WebCore/platform/graphics/TextRun.h: Added textAsString.
There's one new call site that takes advantage of taking the string out.
We could also just consider changing the text funtion to return a
const String&, but to do that safely we'd have to make sure no callers
are using substring or similar functions.

* Source/WebCore/platform/graphics/gstreamer/eme/GStreamerEMEUtilities.h:
(WebCore::GStreamerEMEUtilities::uuidToKeySystem): Removed the peculiar
use of NeverDestroyed here. Since ASCIILiteral is just a pointer, used
it as a return type instead of const ASCIILiteral&.

* Source/WebCore/platform/graphics/win/FontCacheWin.cpp:
(WebCore::FontCache::systemFallbackForCharacters): Updated for name change
of StringView::getCharacters. Would be more elegant to add a
platform-specific StringView::getWideCharactersWithNullTermination
function since this pattern repeats.
(WebCore::createGDIFont): Ditto.
(WebCore::FontCache::getFontSelectionCapabilitiesInFamily): Ditto.

* Source/WebCore/platform/mediastream/CaptureDevice.h: Added include of
NeverDestroyed.h.

* Source/WebCore/platform/win/ClipboardUtilitiesWin.cpp:
(WebCore::createGlobalData): Updated for name change of
StringView::getCharacters.
* Source/WebCore/platform/win/PasteboardWin.cpp:
(WebCore::fileSystemPathFromURLOrTitle): Ditto.
(WebCore::Pasteboard::writeURLToDataObject): Ditto.
(WebCore::createGlobalImageFileDescriptor): Ditto.

* Source/WebKitLegacy/win/DOMCoreClasses.cpp:
(DOMElement::font): Updated for name change of StringView::getCharacters.
Would be more elegant to add a platform-specific StringView::getWideCharacters
functionsince this pattern repeats.

* Source/WebCore/platform/xr/openxr/OpenXRInstance.cpp: Added include of
NeverDestroyed.h.
* Source/WebKit/Platform/LogInitialization.cpp: Ditto.

* Source/WebKit/Shared/API/c/WKString.cpp:
(WKStringGetCharacters): Updated for name change of
StringView::getCharacters.

* Source/WebKit/UIProcess/Inspector/mac/WebInspectorUIProxyMac.mm:
(WebKit::WebInspectorUIProxy::showSavePanel): Make the String to pass to
base64Decode explicitly from the NSString.

* Source/WebKit/webpushd/PushService.mm:
(WebPushD::base64URLDecode): Added. Makes the String to pass explicitly
from the NSString passed in.
(WebPushD::base64Decode): Ditto.

* Source/WebKitLegacy/win/WebKitGraphics.cpp:
(CenterTruncateStringToWidth): Ditto.
(RightTruncateStringToWidth): Ditto.

* Source/WebKitLegacy/win/WebView.cpp:
(WebView::onIMERequestReconvertString): Ditto.

* Tools/WebKitTestRunner/InjectedBundle/InjectedBundlePage.cpp: Added
include of NeverDestroyed.h.

Canonical link: https://commits.webkit.org/255600@main
